cmd/uzfs_test/zrepl_utest.c

```diff
@@ -0,0 +1,411 @@
+	kthread_t *reader;
+	kthread_t *writer;
+	socklen_t in_len;
+	zvol_io_hdr_t hdr;
+	mgmt_ack_t mgmt_ack;
+	struct sockaddr in_addr;
+	struct sockaddr_in replica_io_addr;
+	worker_args_t writer_args, reader_args;
+
+	io_block_size = 4096;
+	active_size = 0;
+	max_iops = 10000;
+	pool = "testp";
+	ds = "ds0";
+
+	io_sfd = new_fd = sfd = -1;
+	mutex_init(&mtx, NULL, MUTEX_DEFAULT, NULL);
+	cv_init(&cv, NULL, CV_DEFAULT, NULL);
+
+	writer_args.threads_done = &threads_done;
+	writer_args.mtx = &mtx;
+	writer_args.cv = &cv;
+	writer_args.io_block_size = io_block_size;
+	writer_args.active_size = active_size;
+	writer_args.max_iops = max_iops;
+
+	reader_args.threads_done = &threads_done;
+	reader_args.mtx = &mtx;
+	reader_args.cv = &cv;
+	reader_args.io_block_size = io_block_size;
+	reader_args.active_size = active_size;
+	reader_args.max_iops = max_iops;
+
+
+	sfd = create_and_bind(tgt_port);
+	if (sfd == -1) {
+		return;
+	}
+
+	rc = listen(sfd, 10);
+	if (rc == -1) {
+		printf("listen() failed with errno:%d\n", rc);
+		goto exit;
+	}
+
+	printf("Listen was successful\n");
+	in_len = sizeof (in_addr);
+	new_fd = accept(sfd, &in_addr, &in_len);
+	if (new_fd == -1) {
+		printf("Unable to accept\n");
+		goto exit;
+	}
+	printf("Connection accepted from replica successful\n");
+
+	hdr.version = REPLICA_VERSION;
+	hdr.opcode = ZVOL_OPCODE_HANDSHAKE;
+	hdr.len = strlen(ds)+1;
+	printf("Op code sent %d with len:%ld\n", hdr.opcode, hdr.len);
+
+	count = write(new_fd, (void *)&hdr, sizeof (zvol_io_hdr_t));
+	if (count == -1) {
+		printf("During hand shake Write error\n");
+		goto exit;
+	}
+	printf("header has been sent with count %d\n", count);
+
+	count = write(new_fd, ds, hdr.len);
+	if (count == -1) {
+		printf("During name send Write error\n");
+		goto exit;
+	}
+	printf("Volname has been sent with count %d\n", count);
+
+	count = read(new_fd, (void *)&hdr, sizeof (zvol_io_hdr_t));
+	if (count == -1) {
+		printf("During hdr Read error\n");
+		goto exit;
+	}
+	printf("Header has read with count %d\n", count);
+
+	count = read(new_fd, (void *)&mgmt_ack, sizeof (mgmt_ack));
+	if (count == -1) {
+		printf("During mgmt Read error\n");
+		goto exit;
+	}
+
+	printf("Vol name: %s\n", mgmt_ack.volname);
+	printf("IP address: %s\n", mgmt_ack.ip);
+	printf("Port: %d\n", mgmt_ack.port);
+
+	bzero((char *)&replica_io_addr, sizeof (replica_io_addr));
+
+	replica_io_addr.sin_family = AF_INET;
+	replica_io_addr.sin_addr.s_addr = inet_addr(mgmt_ack.ip);
+	replica_io_addr.sin_port = htons(mgmt_ack.port);
+	io_sfd = create_and_bind("10315");
+	if (io_sfd == -1) {
+		printf("IO connection Bind at port 10200 failed\n");
+		goto exit;
+	}
+
+	rc = connect(io_sfd, (struct sockaddr *)&replica_io_addr,
+	    sizeof (replica_io_addr));
+	if (rc == -1) {
+		printf("Failed to connect to istgt_controller with err\n");
+		goto exit;
+	}
+	printf("Connect to replica IO port is successfully\n");
+
+	writer_args.sfd = reader_args.sfd = io_sfd;
+	writer = zk_thread_create(NULL, 0,
```

Comment:
multiple writer threads will be good

Reply:
What we want to acheive by multiple writers ? If we go with multiple writer, we need to take lock on socket FD to write to ....correct ? If that is the case then I do not think it make sense to have multiple writer threads because after lock it is serialized i.e. as good as single writer thread. Same goes with reader as well.

Reply:
idea is to test the synchronization among multiple uzfs_zvol_workers.. but, I understand the complexity of having multiple threads..
how about sending multiple IOs in a single go?

Reply:
I understand your concerns, let me have more than one IO connection per replica and test it. It should cover synchronization. I think only thing that I need to take of is to have separate ofsets to write to so that during we can verify what we have written.